osu.Game/Screens/Select/BeatmapCarousel.cs

```diff
@@ -558,71 +573,98 @@ protected override void Update()
+                    }
+                }
 
-                // Only add if we're not already part of the content.
-                if (!scrollableContent.Contains(item))
+                // Add those items within the previously found index range that should be displayed.
+                foreach (var item in toDisplay)
                 {
-                    // Makes sure headers are always _below_ items,
-                    // and depth flows downward.
-                    item.Depth = depth;
+                    var panel = setPool.Get(p => p.Item = item);
 
-                    switch (item.LoadState)
-                    {
-                        case LoadState.NotLoaded:
-                            LoadComponentAsync(item);
-                            break;
+                    panel.Depth = item.CarouselYPosition;
+                    panel.Y = item.CarouselYPosition;
 
-                        case LoadState.Loading:
-                            break;
+                    ScrollableContent.Add(panel);
+                }
+            }
 
-                        default:
-                            scrollableContent.Add(item);
-                            break;
-                    }
+            // Finally, if the filtered items have changed, animate drawables to their new locations.
+            // This is common if a selected/collapsed state has changed.
+            if (revalidateItems)
+            {
+                foreach (DrawableCarouselItem panel in ScrollableContent.Children)
+                {
+                    panel.MoveToY(panel.Item.CarouselYPosition, 800, Easing.OutQuint);
                 }
-                else
+            }
+
+            // Update externally controlled state of currently visible items (e.g. x-offset and opacity).
+            // This is a per-frame update on all drawable panels.
+            foreach (DrawableCarouselItem item in ScrollableContent.Children)
+            {
+                updateItem(item);
+
+                if (item is DrawableCarouselBeatmapSet set)
                 {
-                    scrollableContent.ChangeChildDepth(item, depth);
+                    foreach (var diff in set.DrawableBeatmaps)
+                        updateItem(diff, item);
                 }
             }
+        }
 
-            // this is not actually useful right now, but once we have groups may well be.
-            if (notVisibleCount > 50)
-                itemsCache.Invalidate();
+        private readonly CarouselBoundsItem carouselBoundsItem = new CarouselBoundsItem();
+
+        private (int firstIndex, int lastIndex) getDisplayRange()
+        {
+            // Find index range of all items that should be on-screen
+            carouselBoundsItem.CarouselYPosition = visibleUpperBound - distance_offscreen_to_preload;
+            int firstIndex = visibleItems.BinarySearch(carouselBoundsItem);
+            if (firstIndex < 0) firstIndex = ~firstIndex;
```

Comment:
This is resolved in the code just below:

https://github.com/ppy/osu/pull/10494/files/e662dc5342b26ecedd265b55c121a058006e2ad3#diff-0b1dcf0afe266751400b0c7b7256a23274dc0cab9fc8aab278cf2ce804be079cR658-R660

The comment hopefully explains why it needs to be done like this (the set may be visible and have children – offsetting by one means we don't have to account for this on an item level).

Reply:
Sure, but technically speaking there should be no need to add 1 to the last index if so, because the binary search already guarantees that the set whose index is returned has no overlap with the visible range (its top Y coordinate is either greater than or equal than the upper Y bound).

It's sort of a nitpick though. Shouldn't really ever matter I don't think.

Reply:
Yeah, probably can remove the +1 on the last index. I just left it there for safety (and as you say, it doesn't really matter as there's no need for 100% accuracy here).